lib/icinga/clusterevents-check.cpp

```diff
@@ -93,13 +93,93 @@ void ClusterEvents::ExecuteCheckFromQueue(const MessageOrigin::Ptr& origin, cons
 	ApiListener::Ptr listener = ApiListener::GetInstance();
 
 	if (!listener) {
-		Log(LogCritical, "ApiListener", "No instance available.");
+		Log(LogCritical, "ApiListener") << "No instance available.";
 		return;
 	}
 
+	Defer resetExecuteCommandProcessFinishedHandler ([]() {
+		Checkable::ExecuteCommandProcessFinishedHandler = nullptr;
+	});
+
 	if (params->Contains("source")) {
-		Log(LogCritical, "ApiListener", "Not implemented.");
-		return;
+		String uuid = params->Get("source");
+
+		Host::Ptr host = Host::GetByName(params->Get("host"));
+		if (!host) {
+			Log(LogCritical, "ApiListener") << "Host '" << params->Get("host") << "' not found.";
+			return;
+		}
+
+		Checkable::Ptr checkable;
+		if (params->Contains("service"))
+			checkable = host->GetServiceByShortName(params->Get("service"));
+		else
+			checkable = host;
+
+		if (!checkable) {
+			Log(LogCritical, "ApiListener") << "Checkable '" << params->Get("host")
+				<< "!" << params->Get("service") << "' not found.";
+			return;
+		}
+
+		ObjectLock oLock (checkable);
+
+		if (origin->FromZone && !origin->FromZone->CanAccessObject(checkable)) {
+			Log(LogNotice, "ApiListener")
+				<< "Discarding 'ExecuteCheckFromQueue' event for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': Unauthorized access.";
+			return;
+		}
```

Comment:
(Now I noticed it...)

Please have a look below on what exact checkable is being checked w/o the source param and where Icinga 2 gets it from.

Reply:
I'm not sure if I understood correctly, please correct me if I'm wrong.

You suggest dropping the `checkable` in this section completely, in favour of reusing [the existing execution logic](https://github.com/WuerthPhoenix/icinga2/blob/12da369ee188eae12ad1a2fc1dc6289bbb20bf6d/lib/icinga/clusterevents-check.cpp#L211-L254) below? 
Such that the above section containing the checks would completely disappear.

Reply:
I suggest not to lookup the checkable as you're doing it right now, but to reuse/copy that existing logic. Rationale: You likely don't even have the checkables on command endpoints.